File: AuctionServices/src/main/java/com/artauction/auctionservice/Controllers/PostBidController.java
package com.artauction.auctionservice.Controllers;

import com.artauction.auctionservice.Aops.Roles.RequiresRole;
import com.artauction.auctionservice.Dtos.PostBid.PostBidCreationDto;
import com.artauction.auctionservice.Services.Interfaces.PostBidServices;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/PostBid")
@RequiredArgsConstructor
public class PostBidController {
    // Attributes
    private final PostBidServices postBidServices;

    // Get post bids API
    @GetMapping("/GetAllPostBids")
    public ResponseEntity<?> GetAllPostBids(@RequestParam Integer artworkPostId) {
        return ResponseEntity.ok(postBidServices.getAllPostBids(artworkPostId));
    }

    // Get buyer post bids API
    @RequiresRole("Buyer")
    @GetMapping("/GetBuyerPostBids")
    public ResponseEntity<?> GetBuyerPostBids(@RequestParam Integer buyerId) {
        return ResponseEntity.ok(postBidServices.getAllBuyerBids(buyerId));
    }

    // Create post bid API
    @RequiresRole("Buyer")
    @PostMapping("/CreatePostBid")
    public ResponseEntity<?> CreatePostBid(@RequestBody PostBidCreationDto postBidCreationDto) {
        // Check result
        boolean result = postBidServices.createPostBid(postBidCreationDto);
        if (!result)
            return ResponseEntity.badRequest()
                    .body("Could not place bid. Ensure the post is still active.");

        // Return success
        return ResponseEntity.ok("Bid placed successfully. Wait for final results to appear.");
    }

    // Delete post bid API
    @RequiresRole("Buyer")
    @DeleteMapping("/DeletePostBid")
    public ResponseEntity<?> DeletePostBid(@RequestParam Integer artworkPostId,
                                           @RequestParam Integer buyerId) {
        // Check result
        boolean result = postBidServices.deletePostBid(artworkPostId, buyerId);
        if (!result)
            return ResponseEntity.notFound().build();

        // Return success
        return ResponseEntity.ok("Bid deleted successfully.");
    }
}

File: Project/Backend/src/ArtAuction.API/Controllers/PostBidController.cs
using ArtAuction.Application.DTOs.ArtworkPost;
using ArtAuction.Application.Interfaces.Repositories;
using ArtAuction.Application.Interfaces.Services;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace ArtAuction.API.Controllers;

[ApiController]
[Route("[controller]")]
public class PostBidController : ControllerBase
{
    // Attributes
    private readonly IPostBidServices _postBidServices;
    
    // Constructor
    public PostBidController(IPostBidServices postBidServices)
    {
        this._postBidServices = postBidServices;
    }
    
    // Get post bids API
    [HttpGet("GetAllPostBids")]
    public async Task<IActionResult> GetAllPostBids([FromQuery] int artworkPostId)
    {
        // Return post bids
        return Ok(await _postBidServices.GetAllPostBids(artworkPostId));
    }
    
    // Get buyer post bids
    [Authorize(Roles = "Buyer")]
    [HttpGet("GetBuyerPostBids")]
    public async Task<IActionResult> GetBuyerPostBids([FromQuery] int buyerId)
    {
        // Return post bids
        return Ok(await _postBidServices.GetAllBuyerBids(buyerId));
    }
    
    // Create post bid API
    [Authorize(Roles = "Buyer")]
    [HttpPost("CreatePostBid")]
    public async Task<IActionResult> CreatePostBid([FromBody] PostBidCreationDto postBidCreationDto)
    {
        // Check model state
        if (!ModelState.IsValid)
            return BadRequest(ModelState);
        
        // Check stats
        var result = await _postBidServices.CreatePostBid(postBidCreationDto);
        if (!result)
            return BadRequest("Could not place bid. Ensure the post is still active.");
        
        // Return success stats
        return Ok("Bid placed successfully. Wait for final results to appear.");
    }
    
    // Delete post bid API
    [Authorize(Roles = "Buyer")]
    [HttpDelete("DeletePostBid")]
    public async Task<IActionResult> DeletePostBid([FromQuery] int artworkPostId, [FromQuery] int buyerId)
    {
        // Check model state
        if (!ModelState.IsValid)
            return BadRequest(ModelState);
        
        // Check stats
        var result = await _postBidServices.DeletePostBid(artworkPostId, buyerId);
        if (!result)
            return NotFound("Bid not found or could not be deleted.");
        
        // Return success stats
        return Ok("Bid deleted successfully.");
    }
}